From: win7-support-bot <win7@cromite.org>
Date: Sun, 26 Jul 2026 00:00:00 +0000
Subject: WIN7: tp-dav1d / dav1d

Files:
 third_party/dav1d/libdav1d/src/cpu.c

License: GPL-2.0-or-later - https://spdx.org/licenses/GPL-2.0-or-later.html
---
diff --git a/third_party/dav1d/libdav1d/src/cpu.c b/third_party/dav1d/libdav1d/src/cpu.c
index a8c0abb..811a726 100644
--- a/third_party/dav1d/libdav1d/src/cpu.c
+++ b/third_party/dav1d/libdav1d/src/cpu.c
@@ -81,7 +81,21 @@ COLD int dav1d_num_logical_processors(Dav1dContext *const c) {
 #ifdef _WIN32
 #if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
     GROUP_AFFINITY affinity;
-    if (GetThreadGroupAffinity(GetCurrentThread(), &affinity)) {
+    // A pointer to a kernel32 function that exists only on Win7 and higher
+    typedef BOOL (WINAPI *PFnGetThreadGroupAffinity)(HANDLE, PGROUP_AFFINITY);
+    HMODULE hmodThreadGroup = GetModuleHandleW(L"kernel32.dll");
+
+    PFnGetThreadGroupAffinity pfnGetFunction = (PFnGetThreadGroupAffinity) GetProcAddress(hmodThreadGroup,
+                                                                                            "GetThreadGroupAffinity");
+
+    if (pfnGetFunction == NULL)
+    {
+        SYSTEM_INFO system_info;
+        GetNativeSystemInfo(&system_info);
+        return system_info.dwNumberOfProcessors;
+    }
+
+    if (pfnGetFunction(GetCurrentThread(), &affinity)) {
         int num_processors = 1;
         while (affinity.Mask &= affinity.Mask - 1)
             num_processors++;
